File: TP/chapitre2/correction_tp_factory_typescript/src/creature/implementation.ts
import { Creature } from "./abstract";

export class Dragon extends Creature {
  public static readonly WINGSPAN = 10;
  public static readonly ELEMENT = "Fire";

  constructor() {
    super("Dragon");
  }

  getDescription(): string {
    return `A ${Dragon.ELEMENT} ${this.name} with a wingspan of ${Dragon.WINGSPAN} meters.`;
  }
}

export class Phoenix extends Creature {
  public static readonly REGENERATION_CYCLE = 500;

  constructor() {
    super("Phoenix");
  }

  getDescription(): string {
    return `A ${this.name} with a regeneration cycle of ${Phoenix.REGENERATION_CYCLE} days.`;
  }
}

export class Griffin extends Creature {
  public static readonly HEAD_TYPE = "Aigle";
  public static readonly BODY_TYPE = "Lion";

  constructor() {
    super("Griffin");
  }

  getDescription(): string {
    return `A ${this.name} with a ${Griffin.HEAD_TYPE} head and a ${Griffin.BODY_TYPE} body.`;
  }
}

File: TP/chapitre2/correction_tp_factory/src/creature/implementation.py

from src.creature.abstract import Creature


class Dragon(Creature):

    def __init__(self) -> None:
        super().__init__("Dragon")
        self._element = "Fire"
        self._wingspan = 30  # in meters

    def get_description(self) -> str:
        return f"A {self._element} {self.name} with a wingspan of {self._wingspan} meters."
    

class Phoenix(Creature):

    def __init__(self) -> None:
        super().__init__("Phoenix")
        self._regeneration_cycle = 500  # in years

    def get_description(self) -> str:
        return f"A {self.name} with a regeneration cycle of {self._regeneration_cycle} years."
    

class Griffin(Creature):

    def __init__(self) -> None:
        super().__init__("Griffin")
        self._head_type = "aigle"
        self._body_type = "lion"

    def get_description(self) -> str:
        return f"A {self.name} with a {self._head_type} head and a {self._body_type} body."

